### ISSUE TITLE
The inference is slow  when using bert model converted by tf2onnx.

### ISSUE THREAD

**Issue by DunZhang** [onnx/tensorflow-onnx#821]
Hi,
I find a strange phenomenon.  I have two bert onnx model, one is converted by tf2onnx, another is converted by pytorch. However these two model have differerent inference speed (I use ONNXRuntime to do a inference). The model converted by pytorch is faster about 0.05s than model converted by tf2onnx.
Here is the code of converting models:
```
# convert by using pytorch
torch.onnx.export(model, (features['input_ids'], features['input_mask'], features['segment_ids']),
                  r'G:\torch_bert.onnx', verbose=True,
                  input_names=['input_ids', "input_mask", 'segment_ids'])
# convert by using tf2onnx
python -m tf2onnx.convert --opset 11  --fold_const  --inputs input_ids:0,input_mask:0,segment_ids:0 --outputs output/sen_vec:0  --input bert.pb --output bert_pb.onnx

```

Hence, is there any way to improve the inference speed when using onnx model converted by tf2onnx?

**Comment by jignparm** [comment]
>Hence, is there any way to improve the inference speed when using onnx model converted by tf2onnx?

The two source models are different, so the speed delta does not seem surprising. Is there a reason to believe they should have the same speed? Why not use the PyTorch model in your scenario, especially if it's faster -- unless there's some additional capability in the TF model that's not available in the PyTorch model.

**Comment by guschmue** [comment]
assuming they are the same model.
what version of tf2onnx was this ?

Looking at the ops in this model - some of the slicing on the tensors and int32 ops might not translate well to onnx:

Counter({'Mul': 186, 'Add': 185, 'MatMul': 98, 'Reshape': 71, 'Cast': 66, 'Sub': 62, 'Transpose': 61, 'Concat': 55, 'ReduceMean': 50, 'Sqrt': 25, 'Reciprocal': 25, 'Unsqueeze': 18, 'Softmax': 12, 'Pow': 12, 'Tanh': 12, 'Squeeze': 7, 'Slice': 6, 'Shape': 5, 'Gather': 2, 'Tile': 1, 'Identity': 1, 'Split': 1})

Sequences like this might hurt:

  %Shape_1:0 = Shape(%bert/encoder/Reshape_13:0)
  %Shape_1__458:0 = Cast[to = 1](%Shape_1:0)
  %strided_slice_1:0 = Slice[axes = [0], ends = [1], starts = [0]](%Shape_1__458:0)
  %strided_slice_1__459:0 = Squeeze[axes = [0]](%strided_slice_1:0)
  %strided_slice_1__460:0 = Cast[to = 6](%strided_slice_1__459:0)
  %mul:0 = Mul(%strided_slice_1__460:0, %mul/y:0)
  %Reshape/shape_Unsqueeze__465:0 = Unsqueeze[axes = [0]](%mul:0)
  %Reshape/shape_Concat__467:0 = Concat[axis = 0](%Reshape/shape_Unsqueeze__465:0, %const_fold_opt__524)
  %Reshape__468:0 = Cast[to = 7](%Reshape/shape_Concat__467:0)
  %Reshape_1/shape_Unsqueeze__461:0 = Unsqueeze[axes = [0]](%strided_slice_1__460:0)
  %Reshape_1/shape_Concat__464:0 = Concat[axis = 0](%Reshape_1/shape_Unsqueeze__461:0, %const_fold_opt__522, %const_fold_opt__521)

Need to look into the details to tell if we can do something on it.

**Comment by guschmue** [comment]
just noticed my tf/bert model was on opset 8 which did not have Erf ... need to rebuild it.

**Comment by solarflarefx** [comment]
Out of curiousity, would one expect there to be a performance degradation when running inference on the onnx versus the original model?  For example, if you had a tensorflow model and converted to onnx, would it be expected that the onnx model would inference slower using onnxruntime than say a frozen graph pb model using TensorFlow?

**Comment by guschmue** [comment]
Can happen but it is rare, Most of the time there is some reason, like some transpose to nchw that we can't optimize away or axes come as int32 where onnx wants int64 and we need to insert casts or there is some tf op that we composed with multiple onnx ops.
 
We'll definitely look at this - just need to find some time next few days.

**Comment by solarflarefx** [comment]
So I have been experimenting with a few models and looking at them in Netron.  I am noticing that when I convert to ONNX, transposes seem to be added to various parts of the network.  For example, I have converted a simple MNIST model (attached) and you can see that there is an extra transpose between the MaxPool and the Reshape in the ONNX model that is not in the PB model.  This increases quite a bit when the network is larger.  

Is this expected behavior?  In a large network, would not these transposes negatively affect the inference time?

When debugging, is there a good way to assess where the speed bottlenecks are coming from?

[Mnist_model.zip](https://github.com/onnx/tensorflow-onnx/files/4346170/Mnist_model.zip)

**Comment by guschmue** [comment]
because tf is by default nhwc and onnx is nchw we wrap ops like conv2d in transposes that we optimize away at the end. Sometimes we can't get rid of all of them. Those are costly if there are a lot.
A single one like in your model should not hurt. The cast around the Slice is because onnx slice did not support int32 but since opset-10 onnx does support int32 (we push a change to not do the cast for newer opsets).
Everything else looks ok I think.
You can turn on a profile in the onnxruntime session ... that would tell where the runtime goes.

**Comment by solarflarefx** [comment]
Got it, thanks for the explanation.  

So is there anything that can be done if a lot of transposes do appear in the network?  What about using the --inputs-as-nchw flag?

I would ask the same question about the casts around slices.  It seems that in some of my networks there are also squeezes and unsqueezes that happen around the slices.

Your comment regarding opset -10 is interesting because I had generated that particular onnx with opset -11.  Shouldn't I have expected the cast to go away?

**Comment by guschmue** [comment]
--inputs-as-nchw  makes the first transpose go away in most cases but in a model like resnet50 that 1 transpose doesn't make much difference. 
The int32 support slice in opset-10 and newer: that is something we need to change in tf2onnx first. When a new opset comes out we have forgotten that we worked around the missing type support in the past and since nothing breaks we don't notice. Going to fix this for slice.

**Comment by guschmue** [comment]
this should be resolved.
